Add support for RulesBasedSampler (#466)
signalfx/splunk-otel-java: docs/advanced-config.md

| System property          | Environment variable     | Default value  | Support | Description |
| ------------------------ | ------------------------ | -------------- | ------- | ----------- |
| `otel.traces.sampler`    | `OTEL_TRACES_SAMPLER`     | `always_on`    | Stable  | The sampler to use for tracing.	|

Splunk Distribution of OpenTelemetry Java supports all standard samplers as provided by 
[OpenTelemetry Java SDK Autoconfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure#sampler).
In addition, we add the following two samplers.

### internal_root_off
Setting `otel.traces.sampler` to `internal_root_off` will drop all traces with root spans where `spanKind` is `INTERNAL`, `CLIENT` or `PRODUCER`.

> This setting only keeps root spans where `spanKind` is `SERVER` and `CONSUMER`.

Reviewer: I really really think we should *NOT* document this as something for general use.

Author: `internal_root_off` is already present in our documentation. But I will talk to Ivo and Fabrio about this.

Reviewer 2: Any news?

Reviewer 2: Any news?

Reviewer 2: Any news?

Reviewer 2: Any news?

Reviewer 2: Any news?

### rules
This sampler allows to ignore individual endpoints and drop all traces that would originate from them.
For example, the following configuration results in all requests to `/healthcheck` to be excluded from monitoring: 

```shell
export OTEL_TRACES_SAMPLER=rules

> export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;default=parentbased_always_on

Reviewer: My question is this: is this how otel intends to support complex sampling options? If we document this, and use this format, we're going to be stuck with it, even if otel goes a different way.

Author: I don't know of any plans nor opinions of OpenTelemetry for configuring "complex" samplers via environment variable. I don't know if are ready to commit this configuration. We have to give this to our customer who has requested this functionality and ask for their feedback.

Reviewer: How would I "drop" more than one endpoint?

Author: > How would I "drop" more than one endpoint?

Just add several `drop` rules

Reviewer: this should probably be documented here. It definitely wasn't clear to me from reading it.

```
All requests to downstream services that happen as a consequence of calling an excluded endpoint are also excluded.

The value of `OTEL_TRACES_SAMPLER_ARG` is interpreted as a semicolon-separated list of rules. The following types of rules are supported:

- `drop=<value>`: The sampler drops a span if its `http.url` attribute has a substring equal to the provided value.
You can provide as many `drop` rules as you want.
- `default=sampler`: Fallback sampler used if no `drop` rule matched a  given span. Supported fallback samplers are `always_on` and `parentbased_always_on`. Probability samplers such as `traceidratio` are not supported.

> > If several `default` rules are provided, only the last one will be in effect.

Reviewer 3: ```suggestion
> If several `fallback` rules are provided, only the last one will be in effect.
```

| System property          | Environment variable     | Default value  | Support | Description |
| ------------------------ | ------------------------ | -------------- | ------- | ----------- |
| `otel.traces.sampler`    | `OTEL_TRACES_SAMPLER`     | `always_on`    | Stable  | The sampler to use for tracing.	|

Splunk Distribution of OpenTelemetry Java supports all standard samplers as provided by 
[OpenTelemetry Java SDK Autoconfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure#sampler).
In addition, we add the following two samplers.

### internal_root_off
Setting `otel.traces.sampler` to `internal_root_off` will drop all traces with root spans where `spanKind` is `INTERNAL`, `CLIENT` or `PRODUCER`.
This setting only keeps root spans where `spanKind` is `SERVER` and `CONSUMER`.

### rules
This sampler allows to ignore individual endpoints and drop all traces that would originate from them.

> Let's take a look at the following example configuration:
> ```shell
> export OTEL_TRACES_SAMPLER=rules
> export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;default=parentbased_always_on
> ```
> 
> This configuration will result in all requests to `/healthcheck` endpoint to be excluded from monitoring.
> This includes all requests to downstream services that happen during handling this endpoint by the application.

Reviewer 2: ```suggestion
For example, the following configuration results in all requests to `/healthcheck` to be excluded from monitoring: 

```shell
export OTEL_TRACES_SAMPLER=rules
export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;default=parentbased_always_on
```

All requests to downstream services that happen as a consequence of calling an excluded endpoint are also excluded.
```

Reviewer 2: For some reason, the suggestion feature won't let me include the note below the export snippet.

> Formally, the value of `OTEL_TRACES_SAMPLER_ARG` is interpreted as a `;`-separated list of rules.
> Currently, the following two types of rules are supported:
> 
> - `drop=<value>` - sampler will drop a span if its `http.url` attribute has a substring equal to the provided value.
> You can provide as many `drop` rules as you wish.
> - `default=sampler` - fallback sampler that will be used if no `drop` rule matched given span.
> Supported fallback samplers are `always_on` and `parentbased_always_on`.
> Probability samplers such as `traceidratio` are explicitly not supported.
> If several `default` rules are provided, only the last one will be in effect.
>   
> If `OTEL_TRACES_SAMPLER_ARG` is not provided or has en empty value, then no `drop` rules will be configured and `parentbased_always_on`
> sampler will be used as a default one.

Reviewer 2: ```suggestion
The value of `OTEL_TRACES_SAMPLER_ARG` is interpreted as a semicolon-separated list of rules. The following types of rules are supported:

- `drop=<value>`: The sampler drops a span if its `http.url` attribute has a substring equal to the provided value.
You can provide as many `drop` rules as you want.
- `default=sampler`: Fallback sampler used if no `drop` rule matched a  given span. Supported fallback samplers are `always_on` and `parentbased_always_on`. Probability samplers such as `traceidratio` are not supported.

> If several `default` rules are provided, only the last one will be in effect.
  
If `OTEL_TRACES_SAMPLER_ARG` is not provided or has en empty value, no `drop` rules are configured and `parentbased_always_on` sampler is as default.
```
